### PR TITLE
Jimmy/add error reporting

### PR DESCRIPTION
**Pull request by Jimmyj30**
Draft PR for adding error reporting to the user config lookup

To do:

- [x] update requirements.txt
- [ ] add error reporting to user config lookup
  - [ ] update infra config
  - [ ] update this repo 
- [ ] code changes/cleanup

### CODE REVIEW

**Comment by Jimmyj30** [comment]
@ben-z 

> Looks good overall! One thing to test is the end to end deployment pipeline. i.e. does the deployment update automatically with the latest user configs?

For this, I think I would have to draft another PR in the infra config. Would a github actions step for `deploy-kubernetes` be OK, or would I have to figure out a way to only update this user config lookup in our kubernetes cluster?

File: requirements.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
-uvicorn
-fastapi
-pyyaml
-aiosmtplib
-python-dotenv
+uvicorn==0.23.2
```
Comment (ben-z) [inline requirements.txt]: I recently read something about not upper-bounding Python requirements. Shall we do that?

https://iscinumpy.dev/post/bound-version-constraints/

```suggestion
uvicorn>=0.23.2
```

**Comment by ben-z** [comment]
> @ben-z
> 
> > Looks good overall! One thing to test is the end to end deployment pipeline. i.e. does the deployment update automatically with the latest user configs?
> 
> For this, I think I would have to draft another PR in the infra config. Would a github actions step for `deploy-kubernetes` be OK, or would I have to figure out a way to only update this user config lookup in our kubernetes cluster?

Should be fine to provision all Kubernetes resources. We already have a provisioning step for kubernetes. Just need to change the DRY_RUN env variable to be set on non-master branches only: https://github.com/WATonomous/infra-config/blob/311a7988b0a65dd2064b888d9617680b17f651e1/.github/workflows/provision.yml#L439C14-L439C14

**Comment by ben-z** [comment]
PR looks good so far

**Comment by Jimmyj30** [comment]
> Should be fine to provision all Kubernetes resources. We already have a provisioning step for kubernetes. Just need to change the DRY_RUN env variable to be set on non-master branches only: https://github.com/WATonomous/infra-config/blob/311a7988b0a65dd2064b888d9617680b17f651e1/.github/workflows/provision.yml#L439C14-L439C14

@ben-z 
So we would run DRY_RUN for non-master branches, but set DRY_RUN to true for the master branch only? sounds good.

Also, do you know why this "test" pod can't be accessed outside of the node running it?

eg, this works:

```
root@wato2-ubuntu1:/home/j257jian/repos/infra-config# kubectl -n discord-provisioner-bot  exec --stdin --tty bot-deployment-788b8fd5c9-b6v2r  -- /bin/sh
/app # ls
```

but not this:

```
root@wato2-ubuntu1:/home/j257jian/repos/infra-config# kubectl -n user-config-lookup  exec --stdin --t
ty nginx-857dd58db9-twh4d  -- /bin/sh
Error from server: error dialing backend: dial tcp 100.109.63.65:10250: i/o timeout
```

However, running the same command on the node running the pod does work, wondering if that's something I misconfigured, thanks!

**Comment by ben-z** [comment]
> So we would run DRY_RUN for non-master branches, but set DRY_RUN to true for the master branch only? sounds good.

Ye, can just follow the examples in other jobs


> eg, this works:
> 
> ```
> root@wato2-ubuntu1:/home/j257jian/repos/infra-config# kubectl -n discord-provisioner-bot  exec --stdin --tty bot-deployment-788b8fd5c9-b6v2r  -- /bin/sh
> /app # ls
> ```
> 
> but not this:
> 
> ```
> root@wato2-ubuntu1:/home/j257jian/repos/infra-config# kubectl -n user-config-lookup  exec --stdin --t
> ty nginx-857dd58db9-twh4d  -- /bin/sh
> Error from server: error dialing backend: dial tcp 100.109.63.65:10250: i/o timeout
> ```
> 
> However, running the same command on the node running the pod does work, wondering if that's something I misconfigured, thanks!

Not sure why that is happening! Never seen it before 😮. People talk about firewall issues on google. See what is running on port 10250 on the machine that's hosting the pod? Perhaps something else is occupying the port.

**Comment by Jimmyj30** [comment]
> Not sure why that is happening! Never seen it before 😮. People talk about firewall issues on google. See what is running on port 10250 on the machine that's hosting the pod? Perhaps something else is occupying the port.

Seems to only be pods on tr-ubuntu1 (running commands from wato2-ubuntu1)?

~~seems like tr-ubuntu1 resolves to a weird IP~~

nevermind, seems to be tailscale related actually 

```
j257jian@wato2-ubuntu1:~/repos/infra-config$ ping tr-ubuntu1
PING tr-ubuntu1.stegosaurus-iwato.ts.net (100.109.63.65) 56(84) bytes of data.
```

a quick and weird workaround is to edit the `/etc/hosts` file so we know "tr-ubuntu1" refers to the cluster ip of tr-ubuntu1, or just don't use wato2-ubuntu1 for kubectl stuff ATM.

```
10.0.50.105 tr-ubuntu1.watocluster.local tr-ubuntu1
```

**Comment by ben-z** [comment]
> > Not sure why that is happening! Never seen it before 😮. People talk about firewall issues on google. See what is running on port 10250 on the machine that's hosting the pod? Perhaps something else is occupying the port.
> 
> Seems to only be pods on tr-ubuntu1 (running commands from wato2-ubuntu1)?
> 
> ~seems like tr-ubuntu1 resolves to a weird IP~
> 
> nevermind, seems to be tailscale related actually
> 
> ```
> j257jian@wato2-ubuntu1:~/repos/infra-config$ ping tr-ubuntu1
> PING tr-ubuntu1.stegosaurus-iwato.ts.net (100.109.63.65) 56(84) bytes of data.
> ```
> 
> a quick and weird workaround is to edit the `/etc/hosts` file so we know "tr-ubuntu1" refers to the cluster ip of tr-ubuntu1, or just don't use wato2-ubuntu1 for kubectl stuff ATM.
> 
> ```
> 10.0.50.105 tr-ubuntu1.watocluster.local tr-ubuntu1
> ```

Oh wow good investigation. We have Tailscale installed on most machines, I wonder if it’s happening in those machines as well. Also weird that microk8s is relying on the non FQDN domain name…

**Comment by Jimmyj30** [comment]
quick FYI—secrets are mounted to the container with some symbolic link trickery and such an implementation may need to be accounted for:

https://stackoverflow.com/questions/62776362/k8s-configmap-mounted-inside-symbolic-link-to-data-directory

**Comment by ben-z** [comment]
> quick FYI—secrets are mounted to the container with some symbolic link trickery and such an implementation may need to be accounted for:
> 
> https://stackoverflow.com/questions/62776362/k8s-configmap-mounted-inside-symbolic-link-to-data-directory

Interesting! So pods don't recreate when secrets and config maps change. Good to know!

**Pull request merged**